### ISSUE TITLE
[Process][DotEnv] Passing --env=test through Process does not work

### ISSUE THREAD

**Issue by mhujer** [symfony/symfony#29722]
**Symfony version(s) affected**: 4.2.1

**Description**  
Running `php bin/console --version --env=test` displays that the Symfony runs in `test` env:

```
Symfony 4.2.1 (env: test, debug: true)
```

However, running same command from another Symfony application using Process somehow breaks the `--env=test` and the output is `Symfony 4.2.1 (env: dev, debug: true)`

**How to reproduce**  
I've created a reproducer: https://github.com/mhujer/symfony-envs-reproducer

1) run `composer install` in both `dest` and `process`
2) `cd process` and run `php bin\console app:demo` ([See the code](https://github.com/mhujer/symfony-envs-reproducer/blob/master/process/src/DemoCommand.php#L17))

It runs the console from `dest` application, once with `--env=dev` and once with `--env=test` and you can see that the env is not set correctly for the second run.

```
> php bin\console app:demo
ENV: array(0) {
}
argv: array(3) {
  [0]=>
  string(11) "bin/console"
  [1]=>
  string(9) "--version"
  [2]=>
  string(9) "--env=dev"
}
Symfony 4.2.1 (env: dev, debug: true)
ENV: array(0) {
}
argv: array(3) {
  [0]=>
  string(11) "bin/console"
  [1]=>
  string(9) "--version"
  [2]=>
  string(10) "--env=test"
}
Symfony 4.2.1 (env: dev, debug: true)
```
([I've added two var_dumps](https://github.com/mhujer/symfony-envs-reproducer/commit/d8238440993eb85af4422486cad04a38306b4368#diff-8a35dac6da657887b56d359105bc8972R11) to the other bin/console to see that the parameters are passed correctly.

If I run the same commands from CLI in the `dest` directory, it works as expected:

```
> php bin/console --version --env=dev
ENV: array(0) {
}
argv: array(3) {
  [0]=>
  string(11) "bin/console"
  [1]=>
  string(9) "--version"
  [2]=>
  string(9) "--env=dev"
}
Symfony 4.2.1 (env: dev, debug: true)

>  php bin/console --version --env=test
ENV: array(0) {
}
argv: array(3) {
  [0]=>
  string(11) "bin/console"
  [1]=>
  string(9) "--version"
  [2]=>
  string(10) "--env=test"
}
Symfony 4.2.1 (env: test, debug: true)

```

I tried passing `[]` as a third parameter (`$env`) when creating a new `Process` instance, but it did not help (and the var_dump shows that nothing is in the `$_ENV` anyway).

**Additional context**  
Running on PHP 7.2.14, W10 Pro 64bit

**Comment by xabbuh** [comment]
Does it make a difference when you pass the command arguments as an array?

Like this:

```php
new Process(['php'], ['bin/console'], ['--version'], ['--env=test']);
```

**Comment by mhujer** [comment]
No, it still behaves the same way https://github.com/mhujer/symfony-envs-reproducer/commit/278b46a520d5333867ddcb7f61f8b6f8b64c4075

**Comment by ogizanagi** [comment]
Did you try setting the `Process::__construct()` $env argument to an empty array, so main process env vars aren't forwarded to the process?

**Comment by mhujer** [comment]
@ogizanagi yes, I tried that.

I was able to simplify the repro-case to this code. You can just paste it to existing project (requires `symfony/process`) and see how it works. 


```php
(new \Symfony\Component\Process\Process(['php', 'bin/console', '--version', '--env=dev'], null, []))
	->mustRun(function ($type, $buffer) {
		echo $buffer;
	});
(new \Symfony\Component\Process\Process(['php', 'bin/console', '--version', '--env=test'], null, []))
	->mustRun(function ($type, $buffer) {
		echo $buffer;
	});
```

1. I tried it with a new application created from `symfony/skeleton` and it displays "dev" both times.
2. I tried it with an older application which I have updated to [new "env" structure](https://symfony.com/blog/improvements-to-the-handling-of-env-files-for-all-symfony-versions). It displays "dev" both times
3. But when I tried it with an older application (created using recipes in ~April 2018, SF 4.1), it works fine (and displays "dev" first time and "test" the second time)

**Comment by xabbuh** [comment]
Thanks for the reproducing example. In fact, the following will make your code work:

```php
(new Process(['php', 'bin/console', '--version', '--env=test'], null, ['APP_ENV' => false, 'SYMFONY_DOTENV_VARS' => false]))
            ->mustRun(function ($type, $buffer) {
                echo $buffer;
            });
```

The reason for this is the following: Starting with Symfony 4, environment variables will always be inherited to the child process if you not explicitly pass them with `false` as a value. That's why the `APP_ENV` argument is necessary. The `SYMFONY_DOT_ENV_VARS` part is necessary as this is where the Dotenv component keeps track of already populated environment variables, but for the child process you want to reinitialise this list.

I am not sure if and how we could improve the latter part (the first part is expected).

**Comment by mhujer** [comment]
@xabbuh Thanks a lot, I can confirm it works as expected when explicitly passing false for the variables when creating a process 👍 

However I'm not sure that it was caused by some change in Symfony 4. I just tested it on one application that is running 4.2. It works when I revert the commit where I upgraded it to new `.env` handling according to https://symfony.com/doc/current/configuration/dot-env-changes.html#updating-my-application. So maybe it is related to the recipe? (see my previous comment https://github.com/symfony/symfony/issues/29722#issuecomment-450489175)

**Comment by mhujer** [comment]
@nicolas-grekas It still behaves the same way on new application created from skeleton.

If the environment variables inheritance is expected (as @xabbuh mentioned), I suggest that we add a note to https://symfony.com/doc/current/components/process.html

Something like:

> **Caution:** Environment variables are always inherited to child process unless you pass `['APP_ENV' => false, 'SYMFONY_DOTENV_VARS' => false]` when creating new instance of `Symfony\Component\Process\Process`.

What do you think?

**Comment by xabbuh** [comment]
Improving/clarifying the docs might indeed be a good idea. Can you open an issue in the documentation repository?

**Comment by mhujer** [comment]
OK, created it https://github.com/symfony/symfony-docs/issues/10967

**Comment by treetop1500** [comment]
I am struggling with this issue as well in Symfony 5.1 and have followed the docs and this thread's info to eliminate the inheritance of Env Vars in processes.

I currently have the following as part of a symfony command...

```
$process = new Process(
    ['bin/console','doctrine:database:drop','--force','--env=test'], 
    null, 
    ['APP_ENV' => false, 'SYMFONY_DOTENV_VARS' => false]);
    $process->mustRun(function ($type, $buffer) {
      echo $buffer;
    });
```
When this runs within the command it always drops the dev database, not the test database.

If I run the command from the terminal, it correctly drops the test database.

Any thoughts?

**Comment by stof** [comment]
@treetop1500 env variables should be strings, not booleans

**Comment by treetop1500** [comment]
@stof thank you.

I'm going off the docs from this page... https://symfony.com/doc/current/components/process.html#setting-environment-variables-for-processes

which states:

_In addition to the env vars passed explicitly, processes inherit all the env vars defined in your system. You can prevent this by setting to false the env vars you want to remove:_
```
$process = new Process(['...'], null, [
    'APP_ENV' => false,
    'SYMFONY_DOTENV_VARS' => false,
]);
```

**Comment by treetop1500** [comment]
I also just tried to run this in the command...

```
$process = new Process(
    ['bin/console','doctrine:database:drop','--force','--env=test'], 
    null, 
    ['APP_ENV' => 'test', 'SYMFONY_DOTENV_VARS' => false]);
    $process->run(null, ['APP_ENV' => 'test', 'SYMFONY_DOTENV_VARS' => false]);
```

... with the same result, it drops the dev database.

**Comment by Jo-Chris** [comment]
Still having the same issue on Symfony 6.0

`$process = new Process(['symfony', 'console', 'doctrine:database:create', '--env=test'], null, ['APP_ENV' => false, 'SYMFONY_DOTENV_VARS' => false]);`

`process->run();`

creates the dev-database instead of the testing

// edit

just created a new fresh project to test it there and it still does not work.

**Comment by McLone** [comment]
@Jo-Chris A little late, but I just stomble on the same problem.

Adding false to each var that you have in your .env.test file might work. In your specific case, just adding DATABASE_URL should work :
```php
$process = new Process(
    ['symfony', 'console', 'doctrine:database:create', '--env=test'], 
    null, 
    ['APP_ENV' => false, 'SYMFONY_DOTENV_VARS' => false, 'DATABASE_URL' => false]
);

process->run();
```

It worked for me.
